packages/contracts/utils/etherscan.ts

```diff
@@ -77,3 +86,410 @@ export const runTaskWithRetry = async (
+    ...builtinChains,
+  ].find(({chainId}) => chainId === currentChainId);
+
+  if (!currentChainConfig) {
+    throw new Error(
+      `ChainId ${currentChainId} not supported. Please add the chain details to the hardhat config.`
+    );
+  }
+
+  return currentChainConfig || null;
+};
+
+/**
+ * @returns The Etherscan API key for the current network the hre is connected to.
+ * We look for this key in the hardhat config, but it must be set in the user's `.env` file.
+ */
+export const getApiKey = (
+  hre: HardhatRuntimeEnvironment,
+  chain: ChainConfig
+): string => {
+  // check the apiKey first in the hardhat config, then in the user config
+  const apiKeyByNetwork = (hre.config.etherscan.apiKey ??
+    hre.userConfig.etherscan?.apiKey) as Record<string, string>;
+
+  if (!apiKeyByNetwork) {
+    throw new Error(`Missing Etherscan API keys in the hardhat config.`);
+  }
+
+  if (!(chain.network in apiKeyByNetwork)) {
+    throw new Error(
+      `${chain.network} not supported. Please add the API key to the hardhat config.`
+    );
+  }
+  return apiKeyByNetwork[chain.network];
+};
+
+/**
+ * @returns The API endpoint to hit for the given chain to verify a proxy contract's implementation.
+ */
+export const generateExplorerIsThisAProxyURL = (
+  hre: HardhatRuntimeEnvironment
+): string => {
+  const chain = getChainBlockExplorerDetails(hre);
+  const apiKey = getApiKey(hre, chain);
+
+  const baseUrl = chain.urls.apiURL;
+
+  return (
+    baseUrl + `?module=contract&action=verifyproxycontract&apikey=${apiKey}`
+  );
+};
+
+export const generateVerifyRequest = (
+  url: string,
+  proxy: string,
+  implementation: string
+): AxiosRequestConfig => {
+  return {
+    method: 'POST',
+    headers: {'content-type': 'application/x-www-form-urlencoded'},
+    data: qs.stringify({
+      address: proxy,
+      expectedimplementation: implementation,
+    }),
+    url,
+  };
+};
+
+/**
+ * Sends a message to the block explorer API to verify a proxy contract's implementation.
+ * Logging is printed, and verification errors must be handled by the operator.
+ * @param url the API endpoint that will handle the verification request
+ * @param proxy the details of the proxy contract being verified
+ * @param implementation the details of the implementation contract being verified
+ */
+export const handleLinkProxyRequest = async (
+  url: string,
+  proxy: NameAndAddress,
+  implementation: NameAndAddress | undefined
+): Promise<void> => {
+  if (!implementation) {
+    console.warn(
+      `Implementation not found for ${proxy.name}. Skipping verification.`
+    );
+    return;
+  }
+
+  console.log(
+    `Verifying proxy ${proxy.name} (${proxy.address}) with implementation ${implementation.address}.`
+  );
+
+  const options = generateVerifyRequest(
+    url,
+    proxy.address,
+    implementation.address
+  );
+  const {
+    data: {message, result},
+  } = await axios(options);
```

Comment:
We could use [axios-rate-limit](https://www.npmjs.com/package/axios-rate-limit) or put a simple 6s wait here to not have an additional dependency.

Reply:
Removed Axios